src/utils/queue-object.ts

```diff
@@ -0,0 +1,121 @@
+      if (newState.status === VoiceConnectionStatus.Disconnected) {
+        // Try to reconnect here?
+      } else if (newState.status === VoiceConnectionStatus.Destroyed) {
+        // The connection has been destroyed
+        this.stop();
+      } else if (!this.isConnecting && (newState.status === VoiceConnectionStatus.Connecting || newState.status === VoiceConnectionStatus.Signalling)) {
+        /*
+            In the Signalling or Connecting states, we set a 10 second time limit for the connection to become ready
+            before destroying the voice connection. This stops the voice connection permanently existing in one of these
+            states.
+        */
+        this.isConnecting = true;
+        try {
+          await entersState(this.voiceConnection, VoiceConnectionStatus.Ready, 10000);
+        } catch {
+          if (this.voiceConnection.state.status !== VoiceConnectionStatus.Destroyed)
+            this.voiceConnection.destroy();
+        } finally {
+          this.isConnecting = false;
+        }
+      }
+    });
+
+    // Configure audio player
+    // https://discordjs.github.io/voice/interfaces/audioplayerbufferingstate.html#resource
+    // https://discordjs.github.io/voice/classes/audioresource.html#metadata
+    this.audioPlayer.on('stateChange', (oldState, newState) => {
+      if (newState.status === AudioPlayerStatus.Idle && oldState.status !== AudioPlayerStatus.Idle) {
+        // Previous state is non-idle => Has just finish playing
+        // Non-idle AudioPlayerStatus has a property called `resource` which is the last resource played
+        (oldState.resource as AudioResource<Track>).metadata.onFinish();
+        this.playNext();
+      } else if (newState.status === AudioPlayerStatus.Playing) {
+        // If the Playing state has been entered, then a new track has started playback.
+        (newState.resource as AudioResource<Track>).metadata.onStart();
+      }
+    });
+
+    this.audioPlayer.on('error', (error) =>
+      (error.resource as AudioResource<Track>).metadata.onError(error)
+    );
+
+    this.voiceConnection.subscribe(this.audioPlayer);
+  }
+
+  /**
+   * Adds a new Track to the queue.
+   *
+   * @param track The track to add to the queue
+   */
+  public addTrack(track: Track) {
+    this.songs.push(track);
+    this.playNext();
+  }
+
+  /**
+   * Stops audio playback
+   */
+  public stop() {
+    this.isLocked = true;
+    this.audioPlayer.stop(true);
+  }
+
+  /**
+   * Attempts to play a Track from the queue
+   */
+  private playNext() {
+    // If the queue is locked (already being processed), is empty, or the audio player is already playing something, return
+    if (this.isLocked || this.audioPlayer.state.status !== AudioPlayerStatus.Idle || this.songs.length === 0) {
+      return;
+    }
+    // Lock the queue
+    this.isLocked = true;
+
+    // Take the first item from the queue
+    const nextTrack = this.songs.shift();
```

Comment:
wait we remove the song from the queue to next ? 
Gonna be hard with the loop

Reply:
that's right, maybe we should have a pointer that moves, not shifting the queue :/